include/deal.II/base/enable_ref_counting_by_observer_pointer.h

```diff
@@ -0,0 +1,346 @@
+ * when the object is moved from or destroyed. The mechanism works as follows:
+ * The member function subscribe() accepts a pointer to a boolean that is
+ * modified on invalidation. The object that owns this pointer (usually an
+ * object of class type ObserverPointer) is then expected to check the state of
+ * the boolean before trying to access this class.
+ *
+ * The utility of this class is even enhanced by providing identifying strings
+ * to the functions subscribe() and unsubscribe(). These strings are represented
+ * as <code>const char</code> pointers since the underlying buffer comes from
+ * (and is managed by) the run-time type information system: more exactly, these
+ * pointers are the result the function call <code>typeid(x).name()</code> where
+ * <code>x</code> is some object. Therefore, the pointers provided to
+ * subscribe() and to unsubscribe() must be the same. Strings with equal
+ * contents will not be recognized to be the same. The handling in
+ * ObserverPointer will take care of this.
+ * The current subscribers to this class can be obtained by calling
+ * list_subscribers().
+ *
+ * @ingroup memory
+ */
+class EnableRefCountingByObserverPointer
+{
+public:
+  /**
+   * Constructor setting the counter to zero.
+   */
+  EnableRefCountingByObserverPointer();
+
+  /**
+   * Copy-constructor.
+   *
+   * The counter of the copy is zero, since references point to the original
+   * object.
+   */
+  EnableRefCountingByObserverPointer(
+    const EnableRefCountingByObserverPointer &);
+
+  /**
+   * Move constructor.
+   *
+   * An object inheriting from EnableRefCountingByObserverPointer can only be
+   * moved if no other objects are subscribing to it.
+   */
+  EnableRefCountingByObserverPointer(
+    EnableRefCountingByObserverPointer &&) noexcept;
+
+  /**
+   * Destructor, asserting that the counter is zero.
+   */
+  virtual ~EnableRefCountingByObserverPointer();
+
+  /**
+   * Assignment operator.
+   *
+   * This has to be handled with care, too, because the counter has to remain
+   * the same. It therefore does nothing more than returning <tt>*this</tt>.
+   */
+  EnableRefCountingByObserverPointer &
+  operator=(const EnableRefCountingByObserverPointer &);
+
+  /**
+   * Move assignment operator. Only invalidates the object moved from.
+   */
+  EnableRefCountingByObserverPointer &
+  operator=(EnableRefCountingByObserverPointer &&) noexcept;
+
+  /**
+   * @name EnableRefCountingByObserverPointer functionality
+   *
+   * Classes derived from EnableRefCountingByObserverPointer provide a facility
+   * to subscribe to this object. This is mostly used by the ObserverPointer
+   * class.
+   * @{
+   */
+
+  /**
+   * Subscribes a user of the object by storing the pointer @p validity. The
+   * subscriber may be identified by text supplied as @p identifier.
+   */
+  void
+  subscribe(std::atomic<bool> *const validity,
+            const std::string       &identifier = "") const;
```

Comment:
Related question - are we going to update these names?

Reply:
You mean, change the `subscribe()` function name?

I must admit that I don't care much one way or the other. It might not be a crazy proposition to make all of these member functions `private` and make the `ObserverPointer` a `friend`.